demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,288 @@
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( uint16_t packetIdentifier,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( packetIdentifier != MQTT_PACKET_ID_INVALID );
+
+    size_t listIndex = 0u;
+
+    /* Iterate through record list to find matching topics, and invoke their callbacks. */
+    while( listIndex < recordListCount )
+    {
+        if( matchTopic( pPublishInfo->pTopicName,
+                        pPublishInfo->topicNameLength,
+                        callbackRecordList[ listIndex ].pTopicFilter,
+                        callbackRecordList[ listIndex ].topicFilterLength,
+                        callbackRecordList[ listIndex ].requiresExactMatch ) == true )
+        {
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex++ ].callback( packetIdentifier, pPublishInfo );
+        }
+    }
+}
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           uint16_t topicFileterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) &&
+           ( availableIndex < recordListSize ) )
+    {
+        availableIndex++;
+    }
+
+    if( availableIndex == recordListSize )
+    {
+        /* Record list is full. */
+    }
+    else
+    {
+        /* Should the topic string be copied? */
+        callbackRecordList[ recordListCount ].pTopicFilter = pTopicFilter;
+        callbackRecordList[ recordListCount ].topicFilterLength = topicFileterLength;
+        callbackRecordList[ recordListCount ].callback = callback;
+
+        recordAdded = true;
+
+        /* Update the max index tracker if a higher index spot is found. */
+        if( availableIndex > maxRecordIndex )
+        {
+            maxRecordIndex = availableIndex;
+        }
+    }
+
+    return recordAdded;
+}
+
+void SubscriptionManager_RemoveCallbackRecord( const char * pTopicFilter,
+                                               uint16_t topicFileterLength )
+{
+    size_t matchingRecordIndex = 0u;
+    bool recordFound = false;
+    SubscriptionManager_Record_t * pRecord = NULL;
+
+    /* Iterate through the records list to find the matching record. */
+    do
+    {
+        pRecord = &callbackRecordList[ matchingRecordIndex ];
+
+        if( ( topicFileterLength == pRecord->pTopicFilter ) &&
+            ( strncmp( pTopicFilter, pRecord->pTopicFilter, topicFileterLength ) == 0 ) )
+        {
+            recordFound = true;
+        }
```

Comment:
possible Infinite loop. matchingRecordIndex needs to be incremented.

Reply:
Good catch, fixed!